src/IdentityModel/Client/AccessTokenHandler.cs

```diff
@@ -0,0 +1,195 @@
+﻿// Copyright (c) Brock Allen & Dominick Baier. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+using IdentityModel.Internal;
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace IdentityModel.Client
+{
+    /// <summary>
+    /// HTTP message handler that encapsulates access token handling and renew
+    /// </summary>
+    public class AccessTokenHandler : DelegatingHandler
+    {
+        private readonly SemaphoreSlim _lock = new SemaphoreSlim(1, 1);
+        private readonly TokenClient _tokenClient;
+        private readonly string _scope;
+
+        private string _accessToken;
+        private bool _disposed;
+
+        /// <summary>
+        /// Gets or sets the timeout
+        /// </summary>
+        public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(5);
+        
+        /// <summary>
+        /// Gets the current access token
+        /// </summary>
+        public string AccessToken
+        {
+            get
+            {
+                if (_lock.Wait(Timeout))
+                {
+                    try
+                    {
+                        return _accessToken;
+                    }
+                    finally
+                    {
+                        _lock.Release();
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Occurs when the tokens were renewed successfully
+        /// </summary>
+        public event EventHandler<TokenRenewedEventArgs> TokenRenewed = delegate { };
```

Comment:
Renamed this from `TokenRefreshed` to `TokenRenewed` . I think it will reduce confusion as "refresh" has a specific meaning wrt refresh tokens.